### ISSUE TITLE
Aws AppSync subscriptions not working

### ISSUE THREAD

**Issue by nikolamilanko** [apollographql/apollo-kotlin#2884]
**Summary**
Not able to receive response from the server

**Version**
2.5.2

**Description**
I'm trying to subscribe to AWS AppSync. I followed the example provided in the release not here https://github.com/apollographql/apollo-android/releases. 
The problem is that I'm getting onConnected event but only that, onResponse is never received. Only way to get it work is to put breakpoint on 'socket.send(message.serializeToJson())' within the WebSocketSubscriptionTransport.send method and run it in debug mode. That way OperationClientMessage.Init and OperationClientMessage.Start are delayed a bit (so is switching from SubscriptionManagerState.CONNECTED to SubscriptionManagerState.ACTIVE state ) and subscription is working.

Here is my code 

val okHttpClient = OkHttpClient.Builder()
            .addInterceptor(AuthorizationInterceptor())
            .connectTimeout(30, TimeUnit.MINUTES)
            .build()

       val authorization = mapOf(
            "host" to "***",
            "x-api-key" to "***"
        )

        val webSocketUrl = AppSyncOperationMessageSerializer.buildWebSocketUrl(
            baseWebSocketUrl = "***",
            authorization = authorization
        )

        val serializer = AppSyncOperationMessageSerializer(authorization)

        val subscriptionTransportFactory = WebSocketSubscriptionTransport.Factory(
            webSocketUrl = webSocketUrl,
            webSocketConnectionFactory = okHttpClient,
            serializer = serializer
        )
        val apolloClient = ApolloClient.builder()
            .subscriptionTransportFactory(subscriptionTransportFactory)
            .serverUrl("***")
            .build()

            val test =   apolloClient.subscribe(TestSubscription("***"))

            val callback =
                object : ApolloSubscriptionCall.Callback<TestSubscription.Data> {

                    override fun onResponse(response: Response<TestSubscription.Data>) {
                        Log.e("ApolloSubscriptionCall.Callback", "onResponse $response")
                    }

                    override fun onFailure(e: ApolloException) {
                        Log.e("ApolloSubscriptionCall.Callback", "onFailure $e")
                    }

                    override fun onCompleted() {
                        Log.e("ApolloSubscriptionCall.Callback", "onCompleted")
                    }

                    override fun onTerminated() {
                        Log.e("ApolloSubscriptionCall.Callback", "onTerminated")
                    }

                    override fun onConnected() {
                        Log.e("ApolloSubscriptionCall.Callback", "onConnected")
                    }
                }
            test.execute(callback)
           
And also I found out that adding this part of code with Thread.sleep before subscribing helps in most of the cases.
```
apolloClient.addOnSubscriptionManagerStateChangeListener { fromState, toState ->
            if (fromState == SubscriptionManagerState.CONNECTED && toState == SubscriptionManagerState.ACTIVE) {
                Thread.sleep(1000)
            }
        }
        apolloClient.enableSubscriptions()
        val test = apolloClient.subscribe(TestSubscription("***"))
        apolloClient.subscriptionManager.start()
```

**Comment by vlazdra** [comment]
@martinbonnin Sorry for pinging you, would you maybe have some time to check this issue?

**Comment by martinbonnin** [comment]
Hi @vlazdra  👋 ! Sorry I don't know much about AppSync. This timing issue looks super weird. Does it work sometimes or has it never worked?

**Comment by nikolamilanko** [comment]
It works when I run app in debug mode and delay OperationClientMessage.Start from OperationClientMessage.Init message, otherwise its not working, so my guess is it has something to do with timing of these messages. 
Also I found workaround with this Thread.Sleep block, it works in most of the cases but its not 100% consistent.

**Comment by martinbonnin** [comment]
Hi 👋  Small update for AppSync, the code has been considerably streamlined in the upcoming 3.0 version. Can you try with version `3.0.0-rc03` [migration guide](https://www.apollographql.com/docs/android/migration/3.0/)

**Comment by martinbonnin** [comment]
`3.0.0` has been released and contains a bunch of AppSync improvements, closing this issue. Feel free to leave a message if it's still happening.

**Comment by mboyd1993** [comment]
I'm using Apollo 3.0.0 in a multiplatform project to connect to a GraphQL server hosted on AWS.  I've been able to perform queries and mutations without a problem, but I can't seem to get subscriptions setup with AppSync.  Is there a working example anywhere of how to set it up?  From what I can see in my implementation, it looks like the ApolloClient is never performing the correct handshake with AWS to open the socket.  

Here's what I'm doing to configure the client.  The `serverAddress` and `remoteRealTimeAddress` are the https: and wss: endpoints.

`private val appSyncWsProtocol = AppSyncWsProtocol.Factory(
        mapOf("Authorization" to authToken,
              "host" to serverAddress), 50000)

    // https://docs.aws.amazon.com/appsync/latest/devguide/real-time-websocket-client.html
    private val apolloClient = ApolloClient.Builder()
        .serverUrl(serverAddress)
        .interceptors(listOf(AuthorizationInterceptor(authToken)))
        .webSocketServerUrl(remoteRealTimeAddress)
        .wsProtocol(appSyncWsProtocol)
        .build()`

When I look at the source for the `AppSyncWsProtocol` class, I can see a function called `buildUrl()` that clearly creates the necessary URL for performing the handshake according to the official AppSync documentation (https://docs.aws.amazon.com/appsync/latest/devguide/real-time-websocket-client.html), but it doesn't seem like that function is ever called anywhere.  And indeed, the error message I get when I try to subscribe to anything is this:

`com.apollographql.apollo3.exception.ApolloNetworkException: Connection error:
    {payload={errors=[{message=Both, the "header", and the "payload" query string parameters are missing, errorCode=400}]}, type=connection_error}`

which seems to indicate that the URL being used for the handshake doesn't contain the necessary "header" and "payload" parameters.

Am I doing something wrong?

**Comment by martinbonnin** [comment]
Hi 👋 

There's a test [there](https://github.com/apollographql/apollo-kotlin/blob/d0f29bff0f4c576e54e516f5c701114fd0681c44/tests/websockets/src/commonTest/kotlin/AppSyncTest.kt#L52)

Usually, you need to call `buildUrl` yourself:

```kotlin
    val apiKey = TODO("changeMe")
    val host = "6l5lltvi6fgmrpx5abfxrtq6wu.appsync-api.eu-west-3.amazonaws.com"

    val authorization = mapOf(
        "host" to host,
        "x-api-key" to apiKey
    )
    val url = AppSyncWsProtocol.buildUrl(
        baseUrl = "https://6l5lltvi6fgmrpx5abfxrtq6wu.appsync-realtime-api.eu-west-3.amazonaws.com/",
        authorization = authorization
    )
    val apolloClient = ApolloClient.Builder().networkTransport(
        networkTransport = WebSocketNetworkTransport.Builder().serverUrl(
            serverUrl = url,
        ).protocol(
            protocolFactory = AppSyncWsProtocol.Factory(
                authorization = authorization
            )
        ).build()
    ).build()

```

**Comment by mboyd1993** [comment]
Thank you very much!  I got it working now.  I didn't understand that I needed to create that url using the `buildUrl` method and then set that as the `serverUrl` for the `networkTransport`.  I was incorrectly using the realtime endpoint as the `serverUrl`.  I also changed the `networkTransport` to `subscriptionNetworkTransport` since I'm also using the `ApolloClient` to perform queries and mutations in my implementation.
